Upgrade Ruby to 2.4.5

jholton opened the pull request:
## Description of change
See [issue](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/17866) for description of changes.

## Testing done
Specs

## Testing planned
Manual testing, monitoring logs and Sentry in other environments

## Acceptance Criteria (Definition of Done)

#### Unique to this PR
<!-- This would be a good place to include feature flag check item and info, specific dashboards and instrumentation check item and info -->
- [x] Ruby version set to 2.4.6
- [x] All specs pass
- [x] All deprecations resolved
- [x] update `webmock` gem
- [x] update `json-schema` gem
- [x] update `sidekiq-scheduler` gem, which updates the `rufus-scheduler` gem

#### Applies to all PRs

- [ ] Appropriate logging
- [ ] Swagger docs have been updated, if applicable
- [ ] Provide link to originating GitHub issue, or connected to it via ZenHub
- [ ] Does not contain any sensitive information (i.e. PII/credentials/internal URLs/etc., in logging, hardcoded, or in specs)
- [ ] Provide which alerts would indicate a problem with this functionality (if applicable)

jholton commented on config/settings.yml:
```diff
@@ -32,13 +32,13 @@ binaries:
   clamdscan: /usr/bin/clamdscan
 
-db_encryption_key: f01ff8ebd1a2b053ad697ae1f0d86adb48ebb708021e4c76c3807d37f6b4e389d5aa45ea171f2f5074222784c1ee2bb8272390d1b9517a7a6987c22733ef00b2
+db_encryption_key: f01ff8ebd1a2b053ad697ae1f0d86adb
```
In Ruby 2.3, OpenSSL would just truncate this key to 32 bytes.
In Ruby 2.4, OpenSSL no longer truncates, but now raises an error because it expects a 32 byte key.
[Reference](https://github.com/ruby/openssl/issues/116#issuecomment-286115710)

omgitsbillryan replied:
Assuming this change is backward compatible with instances running Ruby 2.3 (which I think it is?), might be wise to PR & merge on it's own, preemptively to this PR upgrading ruby, to reduce risk.

This will require updates to `credstash` as well, which again, is probably wise to do in advance of the ruby upgrade if we can.

jholton replied:
Yes, it is backward compatible with Ruby 2.3. 
* All my specs pass with the truncated keys
* I did manually:
  * Create a db object that uses the `db_encryption_key` in a Ruby 2.3/long key environment.
  * Then I switched branches to the Ruby 2.4/short key environment and verified that I could retrieve the previously saved data.

These specific changes are only used in local dev/testing, with the other environments values being held in `credstash`, as you mentioned.

I'll open a separate PR for these truncations, and perhaps a devops engineer can truncate the values for `dev` and/or `staging` for now for additional manual testing.

jholton commented on config/settings.yml:
```diff
@@ -32,13 +32,13 @@ binaries:
 relative_url_root: /
 
-secret_key_base: 8af0fe1e378586520e4324694897eb269bd0fffa1c5be6cc3b4ffb9dbde095d0bef5c7fdab73cd05685d8fe1dd589287d78b38e4de7116fbe14461e414072677
+secret_key_base: 8af0fe1e378586520e4324694897eb26
```
Truncated for same reason mentioned above.

jholton commented:
As mentioned in the issue, this will of course require coordinating some changes with devops.  Namely,
- [ ] devops changes to Ruby 2.4.6 for deployed environments
- [ ] devops changes to `secret_key_base` and `db_encryption_key`
- [ ] verify that `sidekiq-scheduler` continues to schedule jobs when expected (due to update of the library).

annaswims commented:
I think you want to remove `TargetRubyVersion: 2.3` from `.rubocop.yml` as well.[ It will default ](https://github.com/rubocop-hq/rubocop/blob/6fe5956c177ca369cfaa70bdf748b70020a56bf4/manual/configuration.md#setting-the-target-ruby-version)to the version in .ruby_version

annaswims commented:
`Dockerfile` Should be updated to `2.4.6-slim-stretch`

annaswims commented:
`docs/setup/native.md` should be updated too.

omgitsbillryan commented:
The version of Ruby being installed on the `vets-api` AMI is going to be: `ruby 2.4.5`, specifically:

```
ruby 2.4.5p335 (2018-10-18 revision 65137) [x86_64-linux-gnu]
```

Probably want to update this PR to be `2.4.5` and not `2.4.6`.

jholton commented:
* Merging `master` created conflits in `Gemfile.lock` so, I reset it to what is current in `master` then re-applied my updates.
* updated .ruby-version file to specify 2.4.5
* updated Dockerfile to specify `2.4.5-slim-stretch`
* updated `docs/setup/native.md` to specify Ruby version 2.4.5
* After rubocop changed `=~` operators to `.match?`, had to use safe navigation operator with `.match?` when evaluating `response_headers['content-type']` in middlewares because that value can be `nil`

jholton commented:
@kreek @annaswims @kfrz @lihanli re-requesting review.  See comment above for changes since last review.  Thanks.
